### PR TITLE
Flexible usage formatting

### PR DESCRIPTION
**Pull request by PureCS**
Hello,

This is the basic skeleton of what I want to do with usage formatting.
* I have managed to extract most of the usage formatting to the `UsageFormatter` and its default subclass `DefaultUsageFormatter`.
* I only left behind `JCommander#usage()` to act as a simple way to print usage, without messing with the underlying `UsageFormatter`.
* I allow injecting `UsageFormatter` into `JCommander` through a setter.

I want to do the following before we consider this solution:
* [X] Try to implement a Unix-style usage formatter to test the flexibility of the solution
* [X] Try to restrict the access of members I exposed in `JCommander` to make this work
* [X] Clean up the `DefaultUsageFormatter` code some more
* [x] Test it

One problem is that this is a public API change, so you might choose to keep some of the old `usage` methods in `JCommander` for backwards compatibility sake.

### CODE REVIEW

File: src/main/java/com/beust/jcommander/UnixStyleUsageFormatter.java [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+package com.beust.jcommander;
+
+import java.util.EnumSet;
+import java.util.List;
+
+/**
+ * A unix-style usage formatter.
+ */
+public class UnixStyleUsageFormatter extends DefaultUsageFormatter {
```
Comment (cbeust) [inline src/main/java/com/beust/jcommander/UnixStyleUsageFormatter.java]: Let's start with an interface `IUsageFormatter` that captures the methods that a formatter needs to implement, then let's have `DefaultUsageFormatter` and `UnixStyleUsageFormatter` implement it.

Comment (PureCS) [reply src/main/java/com/beust/jcommander/UnixStyleUsageFormatter.java]: I have addressed this.

Comment (cbeust) [reply src/main/java/com/beust/jcommander/UnixStyleUsageFormatter.java]: Thanks, looking good. I don't see any specific test for this new feature though, think you could write a few?

Comment (PureCS) [reply src/main/java/com/beust/jcommander/UnixStyleUsageFormatter.java]: I'll look into it.

**Comment by PureCS** [comment]
I need to rebase this, then I think we are done.

File: src/main/java/com/beust/jcommander/DefaultUsageFormatter.java [hunk 2]
```diff
@@ -139,31 +139,55 @@ public void appendAllParametersDetails(StringBuilder out, int indentCount, Strin
...
             // First line, command name
             out.append(indent)
                     .append("  ")
                     .append(parameter.required() ? "* " : "  ")
                     .append(pd.getNames())
                     .append("\n");
-            wrapDescription(out, indentCount, s(indentCount) + pd.getDescription());
+
+            if (hasDescription)
+                wrapDescription(out, indentCount, s(indentCount) + description);
```
Comment (cbeust) [inline src/main/java/com/beust/jcommander/DefaultUsageFormatter.java]: Please use braces, even for single line statements.

File: src/main/java/com/beust/jcommander/DefaultUsageFormatter.java [hunk 3]
```diff
@@ -139,31 +139,55 @@ public void appendAllParametersDetails(StringBuilder out, int indentCount, Strin
...
             Object def = pd.getDefault();
 
             if (pd.isDynamicParameter()) {
                 String syntax = "Syntax: " + parameter.names()[0] + "key" + parameter.getAssignment() + "value";
-                out.append(newLineAndIndent(indentCount)).append(syntax);
+
+                if (hasDescription)
```
Comment (cbeust) [inline src/main/java/com/beust/jcommander/DefaultUsageFormatter.java]: Braces.

File: src/main/java/com/beust/jcommander/DefaultUsageFormatter.java [hunk 4]
```diff
@@ -139,31 +139,55 @@ public void appendAllParametersDetails(StringBuilder out, int indentCount, Strin
...
+                    out.append(s(indentCount));
+                out.append(syntax);
             }
 
             if (def != null && !pd.isHelp()) {
                 String displayedDef = Strings.isStringEmpty(def.toString()) ? "<empty string>" : def.toString();
                 String defaultText = "Default: " + (parameter.password() ? "********" : displayedDef);
-                out.append(newLineAndIndent(indentCount)).append(defaultText);
+
+                if (hasDescription)
```
Comment (cbeust) [inline src/main/java/com/beust/jcommander/DefaultUsageFormatter.java]: Braces.

File: src/main/java/com/beust/jcommander/DefaultUsageFormatter.java [hunk 5]
```diff
@@ -139,31 +139,55 @@ public void appendAllParametersDetails(StringBuilder out, int indentCount, Strin
...
             if (type.isEnum()) {
-                String possibleValues = "Possible Values: " + EnumSet.allOf((Class<? extends Enum>) type);
-                out.append(newLineAndIndent(indentCount)).append(possibleValues);
+                String valueList = EnumSet.allOf((Class<? extends Enum>) type).toString();
+                String possibleValues = "Possible Values: " + valueList;
+
+                // Prevent duplicate values list, since it is set as 'Options: [values]' if the description
+                // of an enum field is empty in ParameterDescription#init(..)
+                if (!description.contains("Options: " + valueList)) {
+                    if (hasDescription)
```
Comment (cbeust) [inline src/main/java/com/beust/jcommander/DefaultUsageFormatter.java]: Braces.

File: src/test/java/com/beust/jcommander/DefaultUsageFormatterTest.java [hunk 6]
```diff
@@ -11,6 +11,78 @@
...
+            public TestEnum1 h = TestEnum1.A;
+            @Parameter(names = {"-i"})
+            public TestEnum2 i;
+            @Parameter(names = {"-k"}, description = "sets k")
+            public TestEnum2 k;
+        }
+
+        // setup
+        StringBuilder sb = new StringBuilder();
+        JCommander jc = new JCommander(new ArgsTemplate());
```
Comment (cbeust) [inline src/test/java/com/beust/jcommander/DefaultUsageFormatterTest.java]: Please use the builder for this, I'd like to completely deprecate the constructor.

**Comment by PureCS** [comment]
Addressed your comments.

**Comment by PureCS** [comment]
I think UsageFormatters sub-classes can be improved though, optimally we want them to be injectable through the builder. But they take `JCommander` as a constructor argument so this is currently impossible to utilize.

One way to address this is by introducing a `JCommander` parameter to each of their methods. Do you have any suggestions?

It is also very weird that half of the project uses 2-spaces indentation and the other half 4-spaces, I wonder if we can address this in the future.

**Comment by cbeust** [comment]
No easy solution to your first question: the formatter needs to retrieve quite a bit of information from the `JCommander` instance to do its job.

**Comment by cbeust** [comment]
Oh and thanks for all your work!

It would be great if you could send a PR to update the documentation next, but that's optional.

**Comment by simon04** [comment]
Having `LANG=en_GB.UTF-8`, the test `DefaultUsageFormatterTest#usageWithRequiredArgsAndResourceBundle` fails with:

```
Can't find bundle for base name MessageBundle, locale en_GB
java.util.MissingResourceException: Can't find bundle for base name MessageBundle, locale en_GB
	at java.util.ResourceBundle.throwMissingResourceException(ResourceBundle.java:1564)
	at java.util.ResourceBundle.getBundleImpl(ResourceBundle.java:1387)
	at java.util.ResourceBundle.getBundle(ResourceBundle.java:773)
	at com.beust.jcommander.DefaultUsageFormatterTest.usageWithRequiredArgsAndResourceBundle(DefaultUsageFormatterTest.java:229)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:100)
	at org.testng.internal.Invoker.invokeMethod(Invoker.java:646)
	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:811)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1137)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:129)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:112)
	at org.testng.TestRunner.privateRun(TestRunner.java:749)
	at org.testng.TestRunner.run(TestRunner.java:603)
	at org.testng.SuiteRunner.runTest(SuiteRunner.java:368)
	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:363)
	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:321)
	at org.testng.SuiteRunner.run(SuiteRunner.java:270)
```

**Comment by PureCS** [comment]
I had the same issue as you, if you want the tests to pass, you either have to force set your locale to a supported one (see [test.resources](https://github.com/cbeust/jcommander/tree/master/src/test/resources)), or add it there (copy-paste the `en_US` one and rename it to `en_GB`).

**Comment by simon04** [comment]
In my understanding, the tests should work independent of locale/environment/timezone settings. Otherwise this causes frustration to new/casual developers.

For this specific test, you might want to specify the US locale (since for this the bundle is contained in the test resources; similar to `com.beust.jcommander.DefaultUsageFormatterTest#i18MissingKeyForCommand`):
```java
java.util.ResourceBundle.getBundle("MessageBundle", Locale.US)
```

**Comment by PureCS** [comment]
Agreed. Thank you for your valuable input, it made me rethink my decision to not address this.

**Comment by cbeust** [comment]
Fair point @simon04, we should hardcode the locale to US for the tests.

**Comment by PureCS** [comment]
@simon04 this has been addressed in https://github.com/cbeust/jcommander/pull/410, let me know how you get on. Have a good day.

**Pull request merged**